packages/api-server/api_server/rmf_io/gateway.py

```diff
@@ -4,18 +4,28 @@
 from rmf_door_msgs.msg import DoorState
 from rx.subject import BehaviorSubject, Subject
 
+from ..models import DoorHealth
+
 
 class RmfGateway:
     def __init__(
         self,
-        initial_door_states: Dict[str, DoorState] = None,
-        initial_building_map: Optional[BuildingMap] = None,
     ):
-        self.door_states = Subject()
-        self.current_door_states: Dict[str, DoorState] = initial_door_states or {}
-        self.building_map = BehaviorSubject(initial_building_map)
-
+        # NOTE: the rx type hints don't actually work https://github.com/ReactiveX/RxPY/issues/514
+        self.door_states: Subject[DoorState] = Subject()
+        self.current_door_states: Dict[str, DoorState] = {}
         self.door_states.subscribe(self._update_door_states)
 
+        self.door_health: Subject[DoorHealth] = Subject()
+        self.current_door_health: Dict[str, DoorHealth] = {}
+        self.door_health.subscribe(self._update_door_health)
```

Comment:
QQ here, sorry, just to see if I'm getting it right, Subject() Represents an object that is both an observable sequence as well as an observer. Each notification is broadcasted to all subscribed observers.

and then you're subscribing an observer to the observable sequence. By using a callback.

and then after instantiating the rmf_gateway, you're using it in the `rmf_io` pushing another observer to the observable sequence again `self.rmf_gateway.door_states.subscribe(self._on_door_state)` .

Can I say that the rmf_gateway is a Singleton?

Reply:
> and then after instantiating the rmf_gateway, you're using it in the rmf_io pushing another observer to the observable sequence again self.rmf_gateway.door_states.subscribe(self._on_door_state) .

The observer subscribes to the observable sequence. There are "producers" like `RmfTransport` and `HealthWatchdog` that pushes items into the subject, then there are "consumers" like `rmf_io` that subscribes to these subjects and perform some tasks accordingly.

> Can I say that the rmf_gateway is a Singleton?

It probably will be in practice, but isn't an actual singleton in implementation.